File: baochuquan/375_Guess_Number_Higher_or_Lower_II.cpp
/*
 solution 1: DP
 dp[i][j] = min(dp[i][j], x + max(dp[i][x-1], dp[x+1][j]));
*/
class Solution {
public:
    int min(int a, int b){
        return a < b ? a : b;
    }
    int max(int a, int b){
        return a > b ? a : b;
    }
    int getMoneyAmount(int n) {
        // solution 1
        int res = 0;
        vector<vector<int> > dp(n+1, vector<int>(n+1, 0));
        for(int i = n-1; i > 0; i--){
            for(int j = i+1; j <= n; j++){
                int ans = INT_MAX;
                for(int x = i; x < j; x++){
                    ans = min(ans, x+max(dp[i][x-1], dp[x+1][j]));
                }
                dp[i][j] = ans;
            }
        }

        return dp[1][n];
    }
};

File: ldzhou/375-Guess-Number-Higher-or-Lower-II.py
# 这题竟然是DP, 其实是被惯性思维误导了, 如果仔细想想, 确实应该是DP

class Solution(object):
    def getMoneyAmount(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = [[0] * (n + 1) for i in range(n + 1)]
        self.DP(dp, 1, n)
        return dp[1][n]
        
    def DP(self, dp, i, j): # 表示i到j, 在最坏的情况下, 我们需要付钱最少是多少.
        if i >= j:
            return 0
        if dp[i][j] != 0:
            return dp[i][j]
        else:
            ans = 2200000000
            for k in range(i, j + 1):
                ans = min(ans, k + max(self.DP(dp, i, k - 1), self.DP(dp, k + 1, j)))
            dp[i][j] = ans
            return ans
